### PR TITLE
Solution pack chemistry port

### PR DESCRIPTION
**Pull request by MorganDawe**
Port of the d6 solution pack 'islandora_solution_pack_chemistry'.

### CODE REVIEW

File: islandora_chemistry.module [hunk 1]
```diff
@@ -4,3 +4,128 @@
...
+    'islandora_chemistry_mods_form' => array(
+      'content_model' => 'islandora:sp_chem_CM',
+      'form_name' => 'Chemistry MODS form',
+      'dsid' => 'MODS',
+      'title_field' => array('titleInfo', 'title'),
+      'transform' => 'mods_to_dc.xsl',
+      'template' => FALSE,
+    ),
+    'islandora_chemistry_mads_form' => array(
+      'content_model' => 'islandora:sp_chem_CM',
```
Comment (nigelgbanks) [inline islandora_chemistry.module]: The MADS form is not needed here, as it was associated with the Researcher which we will use entities for:
https://github.com/discoverygarden/islandora_solution_pack_chemistry/blob/6.x/islandora_chem_sp.module#L46

File: islandora_chemistry.module [hunk 2]
```diff
@@ -4,3 +4,128 @@
  * @file
  * Hooks and callbacks for this module.
  */
+
+/**
+ * Implements hook_theme().
+ */
+function islandora_chemistry_editor_theme($existing, $type, $theme, $path) {
```
Comment (jordandukart) [inline islandora_chemistry.module]: Should be `islandora_chemistry`?

File: islandora_chemistry.module [hunk 3]
```diff
@@ -4,3 +4,128 @@
...
+      'form_file' => "$module_path/xml/islandora_chemistry_form_mods.xml",
+    ),
+    'Chemistry MADS form' => array(
+      'form_file' => "$module_path/xml/islandora_chemistry_form_mads.xml",
+    ),
+  );
+}
+
+/**
+ * Implements hook_islandora_content_model_forms_form_associations().
```
Comment (jordandukart) [inline islandora_chemistry.module]: Deprecated hook, granted a lot of the core solution packs are screwing this up as well. Should be: [`hook_islandora_xml_form_builder_form_associations`](https://github.com/Islandora/islandora_xml_forms/blob/7.x/builder/xml_form_builder.api.php#L38-L58).

**Comment by jordandukart** [comment]
The install is missing `hook_install` and `hook_uninstall` to ingest and purge the required objects respectively. Example [here](https://github.com/Islandora/islandora_solution_pack_pdf/blob/7.x/islandora_pdf.install#L8-L24).

File: includes/chemistry_upload.form.inc [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  $upload_size = min((int) ini_get('post_max_size'), (int) ini_get('upload_max_filesize'));
+  $extensions = array('mol xyz pdb cdx c3d chm cif cml imp gam out cub gau gal jdx dx mop in txt');
+  $tn_extensions = array('jpg jpeg png');
+  $form = array();
+  $form['file'] = array(
+    '#title' => t('File'),
+    '#type' => 'managed_file',
+    '#required' => TRUE,
+    '#description' => t('Select file to upload.<br/>Files must be less than <b>@size MB.</b><br/>Allowed file types: <b>@ext.</b>', array('@size' => $upload_size, '@ext' => $extensions[0])),
+    '#default_value' => isset($form_state['values']['files']) ? $form_state['values']['files'] : NULL,
```
Comment (jordandukart) [inline includes/chemistry_upload.form.inc]: This should be `form_state['values']['file']` to match up with your form element.

File: includes/chemistry_upload.form.inc [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      // Assume it's specified in MB.
+      'file_validate_size' => array($upload_size * 1024 * 1024),
+    ),
+  );
+  $form['tn'] = array(
+    '#title' => t('Thumbnail for object'),
+    '#type' => 'managed_file',
+    '#required' => FALSE,
+    '#description' => t('Select text file to upload.<br/>Files must be less than <b>@size MB.</b><br/>Allowed file types: <b>@ext.</b><br />This file is optional.', array('@size' => $upload_size, '@ext' => $tn_extensions[0])),
+    '#default_value' => isset($form_state['values']['files']) ? $form_state['values']['files'] : NULL,
```
Comment (jordandukart) [inline includes/chemistry_upload.form.inc]: This should be `form_state['values']['tn']` to match up with your form element.

File: includes/chemistry_upload.form.inc [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    '#upload_location' => 'temporary://',
+    '#upload_validators' => array(
+      'file_validate_extensions' => $tn_extensions,
+      // Assume it's specified in MB.
+      'file_validate_size' => array($upload_size * 1024 * 1024),
+    ),
+    '#states' => array(
+      'visible' => array(
+        ':input[name="islandora_chemistry_tn_upload"]' => array('checked' => TRUE),
```
Comment (jordandukart) [inline includes/chemistry_upload.form.inc]: This `#states` references an element that doesn't exist in your form?

File: includes/chemistry_upload.form.inc [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+ * Addes the uploaded file into the ingestable objects datastream.
+ *
+ * @param array $form
+ *   The drupal form.
+ * @param array $form_state
+ *   The drupal form state.
+ */
+function islandora_chemistry_upload_form_submit(array $form, array &$form_state) {
+  module_load_include('inc', 'islandora', 'includes/mime_detect');
+  $object = $form_state['islandora']['objects'][0];
```
Comment (jordandukart) [inline includes/chemistry_upload.form.inc]: Should probably replace this with https://github.com/Islandora/islandora/blob/7.x/includes/ingest.form.inc#L839.

File: includes/chemistry_upload.form.inc [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+/**
+ * Addes the uploaded file into the ingestable objects datastream.
+ *
+ * @param array $form
+ *   The drupal form.
+ * @param array $form_state
+ *   The drupal form state.
+ */
+function islandora_chemistry_upload_form_submit(array $form, array &$form_state) {
+  module_load_include('inc', 'islandora', 'includes/mime_detect');
```
Comment (jordandukart) [inline includes/chemistry_upload.form.inc]: Don't appear to be using `MimeDetect` here at all?

File: islandora_chemistry.module [hunk 9]
```diff
@@ -4,3 +4,128 @@
...
+ * Implements hook_islandora_required_objects().
+ */
+function islandora_chemistry_islandora_required_objects(IslandoraTuque $connection) {
+  $module_path = drupal_get_path('module', 'islandora_chemistry');
+  // Chemistry Content Model.
+  $chemistry_model = $connection->repository->constructObject('islandora:sp_chem_CM');
+  $chemistry_model->owner = 'fedoraAdmin';
+  $chemistry_model->label = 'Islandora Chemistry Content Model';
+  $chemistry_model->models = 'fedora-system:ContentModel-3.0';
+  // ISLANDORACM Datastream.
```
Comment (jordandukart) [inline islandora_chemistry.module]: The only reason to include an `ISLANDORACM` datastream in 7 is if we are trying to preserve cases where Islandora 6 and Islandora 7 may co-exist. If this is the case `ISLANDORACM` should be identical to the XML found [here](https://github.com/roblib/islandora_solution_pack_chemistry/blob/master/xml/CHEMCM.xml).

File: islandora_chemistry.module [hunk 10]
```diff
@@ -4,3 +4,128 @@
...
+/**
+ * Implements hook_islandora_xml_form_builder_forms().
+ */
+function islandora_chemistry_islandora_xml_form_builder_forms() {
+  $module_path = drupal_get_path('module', 'islandora_chemistry');
+  return array(
+    'Chemistry MODS form' => array(
+      'form_file' => "$module_path/xml/islandora_chemistry_form_mods.xml",
+    ),
+    'Chemistry MADS form' => array(
```
Comment (jordandukart) [inline islandora_chemistry.module]: As @nigelgbanks  mentioned shouldn't include a MADS form.

File: xml/islandora_chemistry_content_model.xml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0"?>
```
Comment (jordandukart) [inline xml/islandora_chemistry_content_model.xml]: See [comment](https://github.com/discoverygarden/islandora_solution_pack_chemistry/pull/2/files#r12844385) about `ISLANDORACM`.

File: xml/islandora_chemistry_ds_composite_model.xml [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+  <dsTypeModel ID="DC">
+    <form FORMAT_URI="http://www.openarchives.org/OAI/2.0/oai_dc/" MIME="text/xml"/>
+  </dsTypeModel>
+  <dsTypeModel ID="RELS-EXT" optional="true">
+    <form FORMAT_URI="info:fedora/fedora-system:FedoraRELSExt-1.0" MIME="application/rdf+xml"/>
+  </dsTypeModel>
+  <dsTypeModel ID="MODS" optional="true">
+    <form MIME="text/xml"></form>
+  </dsTypeModel>
+  <dsTypeModel ID="MADS" optional="true">
```
Comment (jordandukart) [inline xml/islandora_chemistry_ds_composite_model.xml]: MADS likely not needed as we are removing the form away?

File: xml/islandora_chemistry_form_mads.xml [hunk 13]
```diff
@@ -0,0 +1,505 @@
+<?xml version="1.0"?>
```
Comment (jordandukart) [inline xml/islandora_chemistry_form_mads.xml]: No longer need?

**Comment by willtp87** [comment]
Had a chat with the Jordan and Morgan.  There are some issues with the base module that we are merging into that we should probably take care of here. It may have been a little ambitious. The README has todos that should be done, the build.xml's lint files need updating, https://github.com/discoverygarden/islandora_solution_pack_chemistry/blob/7.x/build.xml#L87-L91 and any unused files/directories should be removed.

**Pull request merged**